modules/http-api/src/main/java/org/mule/runtime/http/api/server/ws/WebSocketConnectionHandler.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.http.api.server.ws;
+
+import org.mule.api.annotation.Experimental;
+import org.mule.runtime.http.api.ws.WebSocket;
+import org.mule.runtime.http.api.ws.WebSocketCloseCode;
+
+/**
+ * Handles connection aspects of an inbound {@link WebSocket}
+ *
+ * @since 4.1.5 as experimental
+ */
+@Experimental
+public interface WebSocketConnectionHandler {
+
+  /**
+   * Generates the id of the socket that will be created from the given {@code request}
+   *
+   * @param request the inbound request to open a {@link WebSocket}
+   * @return an id. Cannot be null or empty.
+   */
+  String getSocketId(WebSocketRequest request);
+
+  /**
+   * Invoked when the {@code socket} is established
+   *
+   * @param socket  the established {@link WebSocket}
+   * @param request the inbound {@link WebSocketRequest request}
+   * @throws WebSocketConnectionRejectedException if the connection is refused by the owning server
+   */
+  void onConnect(WebSocket socket, WebSocketRequest request) throws WebSocketConnectionRejectedException;
```

Comment:
This will be triggered each time a new web socket is created as a result of an incoming request? How does this relate to `getSocketId`? Seems like first, a socket id will be established since that is needed for a `WebSocket` instance so that would actually execute before the `onConnect`?

Reply:
getSocketId is invoked first. This is invoked after.